class.c

```diff
@@ -1349,8 +1350,9 @@ class_descendants_recursive(VALUE klass, VALUE v)
     struct subclass_traverse_data *data = (struct subclass_traverse_data *) v;
 
     if (BUILTIN_TYPE(klass) == T_CLASS && !FL_TEST(klass, FL_SINGLETON)) {
-        if (data->buffer && data->count < data->maxcount) {
-            data->buffer[data->count] = klass;
+        if (data->buffer && data->count < data->maxcount && !rb_objspace_garbage_object_p(klass)) {
```

Comment:
Is `rb_objspace_garbage_object_p()` reliable and will always returns false if `klass` was collected?
i.e. is it not possible the object at that address is now a different Class or worse a different object which is not a class? What prevents from that address to be reused for another object?

Reply:
It is very subtle. If I understand correctly what @ko1 is saying, the allocation may trigger the GC, but the finalizers of collected objects are not immediately executed, but delayed to the next interrupt check point. This means any new objects are not created at this point, so the scenario that you said should not occur. (Also, in fact, `maxcount` check is not needed.)

I don't want to rely on this subtle implementation detail, but I have no idea to fix this issue soon.